packages/playwright-core/src/server/injected/highlight.ts

```diff
@@ -113,90 +114,119 @@ export class Highlight {
     this._actionPointElement.hidden = true;
   }
 
+  clearHighlight() {
+    for (const entry of this._highlightEntries) {
+      entry.highlightElement?.remove();
+      entry.tooltipElement?.remove();
+    }
+    this._highlightEntries = [];
+  }
+
   updateHighlight(elements: Element[], selector: string, isRecording: boolean) {
+    let color: string;
+    if (isRecording)
+      color = '#dc6f6f7f';
+    else
+      color = elements.length > 1 ? '#f6b26b7f' : '#6fa8dc7f';
+    this._innerUpdateHighlight(elements, { color, tooltipText: selector });
+  }
+
+  maskElements(elements: Element[]) {
+    this._innerUpdateHighlight(elements, { color: '#F0F' });
+  }
+
+  private _innerUpdateHighlight(elements: Element[], options: { color: string, tooltipText?: string }) {
     // Code below should trigger one layout and leave with the
     // destroyed layout.
 
-    // Destroy the layout
-    this._tooltipElement.textContent = selector;
-    this._tooltipElement.style.top = '0';
-    this._tooltipElement.style.left = '0';
-    this._tooltipElement.style.display = 'flex';
-
-    // Trigger layout.
-    const boxes = elements.map(e => e.getBoundingClientRect());
-    const tooltipWidth = this._tooltipElement.offsetWidth;
-    const tooltipHeight = this._tooltipElement.offsetHeight;
-    const totalWidth = this._innerGlassPaneElement.offsetWidth;
-    const totalHeight = this._innerGlassPaneElement.offsetHeight;
-
-    // Destroy the layout again.
-    if (boxes.length) {
-      const primaryBox = boxes[0];
-      let anchorLeft = primaryBox.left;
+    if (this._highlightIsUpToDate(elements))
+      return;
+
+    // 1. Destroy the layout
+    this.clearHighlight();
+
+    for (let i = 0; i < elements.length; ++i) {
```

Comment:
Add a separate comment for this step too ?

Reply:
This is the same step!